### PR TITLE
perf_cnt: add the tracing of window-based average calculation

### PR DESCRIPTION
**Pull request by johnylin76**
The average number of cpu cycles/ticks spent could be provided per 1024
scheduled tasks by the performance counter. It would be more accurate for
MCPS measurement than the peak number. Even so, the tracing macro will log
both of them to provide more insights.

This is enabled by CONFIG_PERFORMANCE_COUNTERS_RUN_AVERAGE, which depends
on CONFIG_PERFORMANCE_COUNTERS.

### CODE REVIEW

**Comment by cujomalainey** [comment]
Hi Johnny, please address CI issues then ping again for review

**Comment by johnylin76** [comment]
Only 2 checks left from "checkpatch --strict", are they negligible?

CHECK: Macro argument reuse 'pcd' - possible side-effects?
#59: FILE: src/include/sof/audio/component.h:216:
+#define comp_perf_avg_info(pcd, comp_p)					\
+	comp_info(comp_p, "perf comp_copy cpu avg %u (current peak %u)",\
+		  (uint32_t)((pcd)->cpu_delta_sum),			\
+		  (uint32_t)((pcd)->cpu_delta_peak))

CHECK: Macro argument reuse 'pcd' - possible side-effects?
#109: FILE: src/include/sof/lib/perf_cnt.h:76:
+#define perf_cnt_average(pcd, trace_m, arg) do {                             \
+		(pcd)->cpu_delta_sum += (pcd)->cpu_delta_last;               \
+		if (++(pcd)->sample_cnt == 1 << PERF_CNT_CHECK_WINDOW_SIZE) {\
+			(pcd)->cpu_delta_sum >>= PERF_CNT_CHECK_WINDOW_SIZE; \
+			trace_m(pcd, arg);                                   \
+			(pcd)->cpu_delta_sum = 0;                            \
+			(pcd)->sample_cnt = 0;                               \
+		}                                                            \
+	} while (0)

**Review by lgirdwood — APPROVED** [review]
@kv2019i @johnylin76  we should make sure we have the same performance monitoring with xtos and Zephyr (including strings). This will allow for easier comparisons.

**Comment by lgirdwood** [comment]
> Only 2 checks left from "checkpatch --strict", are they negligible?

This is fine (for GCC), but always safer to prefix if needed to avoid collisions.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by johnylin76** [comment]
> @kv2019i @johnylin76 we should make sure we have the same performance monitoring with xtos and Zephyr (including strings). This will allow for easier comparisons.

IIUC as well as perf_cnt_stamp(), the newly-added perf_cnt_average() doesn't act differently between xtos and Zephyr. (or do I need the additional work on that?)

**Review by kv2019i — APPROVED** [review]
Excellent, thank you @johnylin76 ! And @lgirdwood , we'll get these same traces both on XTOS and Zephyr, so we are good with that.

File: src/include/sof/audio/component.h [hunk 1]
```diff
@@ -209,10 +209,15 @@ enum {
 		      trace_comp_get_subid, comp_p, __e, ##__VA_ARGS__)
 
 #define comp_perf_info(pcd, comp_p)					\
-	comp_info(comp_p, "perf comp_copy peak plat %d cpu %d",		\
+	comp_info(comp_p, "perf comp_copy peak plat %u cpu %u",		\
```
Comment (marc-hb) [inline src/include/sof/audio/component.h:212]: This line looks like a bug fix, would be nice to have it in a separate commit.

**Review by marc-hb — COMMENTED** [review]
Nothing against _adding_ a new metric but keep in mind averages hide latency issues.
https://bravenewgeek.com/everything-you-know-about-latency-is-wrong/

I prefer VoIP calls because of the higher bandwidth and clearer sound but I regularly revert to "POTS" calls because the deterministic lack of drop outs is much more important than voice clarity.

File: src/schedule/ll_schedule.c [hunk 2]
```diff
@@ -83,6 +83,16 @@ static void perf_ll_sched_trace(struct perf_cnt_data *pcd, int ignored)
 		(uint32_t)((pcd)->plat_delta_peak),
 		(uint32_t)((pcd)->cpu_delta_peak));
 }
+
+#if CONFIG_PERFORMANCE_COUNTERS_RUN_AVERAGE
+static void perf_avg_ll_sched_trace(struct perf_cnt_data *pcd, int ignored)
```
[the remark this answers is not in the document]
Comment (johnylin76) [reply src/schedule/ll_schedule.c:88]: what is edf?

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/schedule/ll_schedule.c:88]: Earliest Deadline First - it's not complete in xtos mode and difficult to measure the preemption/utilization with xtos. Measurable with Zephyr though.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@johnylin76 can you check the Zephyr build in CI.

**Comment by marc-hb** [comment]
The Zephyr build failure looks like it should have been fixed by 9e18603636dd532 which was merged 12 days ago. Strange.

`SOF version.cmake` is unfortunately broken right now, probably because of a recent git security fix.

**Comment by marc-hb** [comment]
> SOF version.cmake is unfortunately broken right now,

SOF version.cmake issue fix submitted, fast-tracked and merged (thx @lgirdwood ) in #5730. See SF version.cmake now working correctly in sample run https://github.com/thesofproject/sof/runs/6167251165

@johnylin76 can you please do a simple `git commit --amend -C HEAD` and force-push to restart all CIs?

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
> SOF version.cmake issue fix submitted, fast-tracked and merged (thx @lgirdwood ) in #5730. See SF version.cmake now working correctly in sample run https://github.com/thesofproject/sof/runs/6167251165
> 
> @johnylin76 can you please do a simple `git commit --amend -C HEAD` and force-push to restart all CIs?

Thanks @marc-hb , I thought the CI would merge with HEAD, but maybe not. @johnylin76 we can merge once you can rebase and push (and CI passes).

**Comment by marc-hb** [comment]
No need to rebase (unless there is a git conflict), this is a very common misconception. A fast `git commit --amend -C HEAD` and force-push is enough. The vast majority of CIs test the constantly moving `pull/5680/merge`, they do not test the static `pull/5680/head`. See for instance https://sof-ci.01.org/sofpr/PR5680/build150/build/tgl_gcc.txt (and many others)

```
-- SOF version.cmake starting at 2022-04-27T10:02:48Z UTC
-- /home/sof/work/sof.git is at git commit with parent(s):
commit 47c82f60b7d7ebe1b6c75ac4a6d1edddb786f09e (HEAD, sof/pull/5680/merge)
    d67e2ea8a4bccdadc00971100ddf2ff459dc8fb8
    cd456c60c6bc00e8f6222005ea19e1c980828a61 
Merge: d67e2ea8a cd456c60c
Author: Johny Lin <lpg76627@gmail.com>
Date:   Wed Apr 27 10:57:37 2022 +0100

    Merge cd456c60c6bc00e8f6222005ea19e1c980828a61 into d67e2ea8a4bccdadc00971100ddf2ff459dc8fb8

```

**Comment by johnylin76** [comment]
Sorry for the late reply. I had done "git commit --amend -C HEAD" and force push commits.

**Review by lgirdwood — APPROVED** [review]
